rfcs/0151-nixos-port-alloc.md

```diff
@@ -0,0 +1,135 @@
+
+The less likely used ports of the port space in a system are the higher ones,
+and the highest one is 65535, so a NixOS module could keep track of which services
+need a port and the service modules need only to reference that port in their
+configurations.
+
+This module exposes the options under `networking.ports`. A service module can
+request a port by defining `networking.ports.service.enable = true` and get the
+allocated port by referring to `networking.ports.service.port`. The service doesn't
+depend on which logic the allocator uses to generate the port number. Only asks for
+a port and get the port to be used.
+
+The port allocator will allocate ports decremented from 65535 so it's very unlikely
+that it will reach ports under 1024. Because of the ordered nature of attrsets.
+
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+This is how the module would be used:
+```nix
+{ config, lib, ... }:
+lib.mkIf config.service.foo.enable {
+    networking.ports.foo-web.enable = true;
+    service.foo.port = mkDefault config.networking.ports.foo-web.port;
+}
+```
+
+And an already working implementation of the specification:
+```nix
+{ config, lib, ... }:
+
+let
+  inherit (builtins) removeAttrs;
+  inherit (lib) mkOption types submodule literalExpression mdDoc mkDefault attrNames foldl' mapAttrs mkEnableOption attrValues;
+in
+
+{
+  options.networking.ports = mkOption {
+    default = {};
+
+    example = literalExpression ''{
+      {
+        app.enable = true;
+      }
+    }'';
+
+    description = "Build time port allocations for services that are only used internally";
+
+    apply = ports: lib.pipe ports [
+      (attrNames) # gets only the names of the ports
+      (foldl' (x: y: x // {
+        "${y}" = (ports.${y}) // {
+          port = x._port;
+        };
+        _port = x._port - 1;
+      })  {_port = 65534; }) # gets the count down of the ports
+      (x: removeAttrs x ["_port"]) # removes the utility _port entity
+    ];
+
+    type = types.attrsOf (types.submodule ({ name, config, options, ... }: {
+      options = {
+        enable = mkEnableOption "Enable automatic port allocation for service ${name}";
+        port = mkOption {
+          description = "Allocated port for service ${name}";
+          type = types.nullOr types.port;
+        };
+      };
+    }));
+  };
+
+  config.environment.etc = lib.pipe config.networking.ports [
+    (attrNames)
+    (foldl' (x: y: x // {
+      "ports/${y}" = {
+        inherit (config.networking.ports.${y}) enable;
+        text = toString config.networking.ports.${y}.port;
+      };
+    }) {})
+  ];
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- If someone externally expects to use that service directly, the port which could be used
+ to access may differ like a local IP when it's not reserved by the router so it's not
+ recommended to use this module in these cases.
+
+# Alternatives
+[alternatives]: #alternatives
```

Comment:
To sum up some of the other threads into an alternative proposal.

## Different Modes

The current proposal only deals with automatically assigned ports. This means that it is unaware of static ports on the system. I think it may be better to make the port allocator (maybe better called a "manager" now) know about all ports to make its decision.

I would add two or three modes to the allocator

1. `{ port = 1234 }`. This is *required mode* In this mode the specific port is reserved for this service. It will not be assigned to another service. An error is raised at eval time if two services require the same port.
2. `{}` This is *random mode* In this mode an otherwise non-allocated port is assigned.
3. `{ mode = "prefer"; port = 1234 }` The port will be assigned to one of the *preferred mode* services if not taken by a *required mode* service. If the port is taken by a *required mode* or another *preferred mode* service it is given a port as if it was *random mode*. (IDK if this is necessary, maybe we can skip it for the first pass. The main benefit is aesthetics but it raises the chance of a re-assignment causing a problem if people are relying on the preferred port. It can always be added later.)

Most notably `mode = "require"` allows existing modules to use the port allocator and remain backwards compatible. This provides good error messages on conflict and avoids collisions between these services and randomly assigned ports. I would recommend migrating all existing services to `mode = "require"`.

## Mostly Stable Assignment

I would update the assignment algorithm to be mostly stable. This reduces disruption when changing enabled services. In the vast majority of changes there would be no reassignment. In cases where there would have been a collision only a small number of services (hopefully 1) would change ports.

The most likely method of doing this would be consistent hashing using the port name. On collision a new hash would be run to pick an alternate port.

## Reserved Ranges

I think the allowed or reserved ranges should be customizable. By default we should probably exclude 0-1024 (root-only) and the ephemeral port range.

## Different IPs

It would also be nice to be able to reserve ports separately on different IPs. This is probably a nice-to-have and can be implemented later. If I reserve some ports on `127.0.0.1` it would be nice if I can also use those on my public IP. Maybe we could even assign IP addresses out of a range like `127.0.0.0/8`. But this is probably fairly niche. In most cases assigning ports globally is unlikely to run out and is likely simpler when trying to debug issues on a system as there should be no port reuse so you can mostly ignore IPs.

This would be forwards-compatible to add later. Ports that don't specify an IP are assumed to use `::` (and collide with all ports on all addresses).

```nix
networking.ports.foo = {
  address = "127.0.0.8";
};
```

## Schema

My current thought is the flowing schema:

<dl>
<dt><code>networking.ports.assignmentRange.start</code></dt>
<dd>

The first port number that is eligible for assignment. Defaults to `1`.

</dd>

<dt><code>networking.ports.assignmentRange.end</code></dt>
<dd>

The first port number that is not eligible for assignment. Defaults to `2^16`.

</dd>

<dt><code>networking.ports.reservedRanges</code></dt>
<dd>

List of `{start = 0; end = 1024}` ranges that won't be assigned from. Defaults to `[{start = 0; end = 1024} {start = ephemeral_start; end =- ephemeral_end}`

</dd>

<dt><code>networking.ports.port.*</code></dt>
<dd>

Attrset of ports. Presence indicates that a port should be allocated. (I can't think of a use case to disable, it would likely cause an eval failure for whatever reads the port). This could likely be proxied via `services.*.port` or similar in many cases.

</dd>

<dt><code>networking.ports.port.*.address</code></dt>
<dd>

(Possible future addition). Defaults to `::`.

</dd>

<dt><code>networking.ports.port.*.mode</code></dt>
<dd>

(Possible  future addition). Used if `port` is set. Enum of `"required" or `"preferred". Defaults to `"required"`.

</dd>

<dt><code>networking.ports.port.*.port</code></dt>
<dd>

Defaults to `null`. If set this port is fixed and used only for avoidance for other assignments and raising an error on conflict between manually-assigned ports. This is also set by the allocator so that it can be referenced as `config.networking.ports.port.foo.port` when generating config files for dependent services.

</dd>
</dl>